drivers/usb/device/usb_dc_numaker.c

```diff
@@ -0,0 +1,1943 @@
+
+#define NUMAKER_USBD_EP_MAXNUM USBD_MAX_EP
+
+/* Message type */
+#define NUMAKER_USBD_MSG_TYPE_SW_RECONN 0 /* S/W reconnect */
+#define NUMAKER_USBD_MSG_TYPE_CB_STATE  1 /* Callback for usb_dc_status_code */
+#define NUMAKER_USBD_MSG_TYPE_CB_EP     2 /* Callback for usb_dc_ep_cb_status_code */
+
+/* Additional register definition */
+#if defined(CONFIG_USB_DC_NUMAKER_ALLOW_ISO_IN_OUT_SAME_NUM)
+#define NUMAKER_USBD_ALLOW_ISO_IN_OUT_SAME_NUM_REG 0x001C
+#define NUMAKER_USBD_ALLOW_ISO_IN_OUT_SAME_NUM_Msk 0x2
+#endif
+
+/* Message structure */
+struct numaker_usbd_msg {
+	uint32_t msg_type;
+	union {
+		struct {
+			enum usb_dc_status_code status_code;
+		} cb_state;
+		struct {
+			uint8_t ep_addr;
+			enum usb_dc_ep_cb_status_code status_code;
+		} cb_ep;
+	};
+};
+
+/* Forward declarations */
+struct usb_dc_numaker_device;
+
+/* Immutable device context */
+struct usb_dc_numaker_config {
+	USBD_T *usbd_base;
+	const struct reset_dt_spec reset;
+	uint32_t clk_modidx;
+	uint32_t clk_src;
+	uint32_t clk_div;
+	const struct device *clkctrl_dev;
+	void (*irq_config_func)(struct usb_dc_numaker_device *dev);
+	void (*irq_unconfig_func)(struct usb_dc_numaker_device *dev);
+	const struct pinctrl_dev_config *pincfg;
+	uint32_t num_bidir_endpoints;
+	uint32_t dmabuf_size;
+	int maximum_speed;
+};
+
+/* EP context */
+struct numaker_usbd_ep {
+	bool valid; /* Valid */
+
+	bool nak_clr; /* NAK cleared (ACK next transaction) */
+
+	struct usb_dc_numaker_device *dev; /* Pointer to the containing device */
+
+	uint8_t usbd_hw_ep_hndl; /* BSP USBD driver EP handle EP0, EP1, EP2, etc */
+	uint32_t usbd_hw_ep_cfg; /* BSP USBD driver EP configuration */
+
+	/* EP DMA buffer */
+	bool dmabuf_valid;
+	uint32_t dmabuf_base;
+	uint32_t dmabuf_size;
+
+	/* On USBD, no H/W FIFO. Simulate based on above DMA buffer with
+	 * one-shot implementation
+	 */
+	uint32_t read_fifo_pos;
+	uint32_t read_fifo_used;
+	uint32_t write_fifo_pos;
+	uint32_t write_fifo_free;
+
+	/* NOTE: On USBD, Setup and CTRL OUT are not completely separated. CTRL OUT MXPLD
+	 * can be overridden to 8 by next Setup. To overcome it, we make one copy of CTRL
+	 * OUT MXPLD immediately on its interrupt.
+	 */
+	uint32_t mxpld_ctrlout;
+
+	/* EP address */
+	bool ep_addr_valid;
+	uint8_t ep_addr;
+
+	/* EP MPS */
+	bool ep_mps_valid;
+	uint16_t ep_mps;
+
+	usb_dc_ep_callback cb; /* EP callback function */
+};
+
+/* EP context manager */
+struct numaker_usbd_ep_mgmt {
+	/* EP context management
+	 *
+	 * Allocate-only, and de-allocate all on re-initialize in usb_dc_attach().
+	 */
+	uint8_t ep_idx;
+
+	/* DMA buffer management
+	 *
+	 * Allocate-only, and de-allocate all on re-initialize in usb_dc_attach().
+	 */
+	uint32_t dmabuf_pos;
+
+	/* Pass Setup packet from ISR to thread */
+	bool new_setup;
+	struct usb_setup_packet setup_packet;
+
+	struct numaker_usbd_ep ep_pool[NUMAKER_USBD_EP_MAXNUM];
+};
+
+/* Mutable device context */
+struct usb_dc_numaker_data {
```

Comment:
Drivers data, more consisted would be `struct numaker_usbd_data`

Reply:
here and elsewhere, changed naming from `usb_dc_numaker_` to `numaker_usbd_` for consistent